### ISSUE TITLE
Permissive settings by default?

### ISSUE THREAD

**Issue by attipaci** [nom-tam-fits/nom-tam-fits#161]
I'd like to start a discussion on a proposal, which is this: __The library should be able to read FITS files as best as it can by default__.

Or more specifically, I'd like the library to be able to read FITS out of the box, which may be:

- using supported conventions (e.g. HIERARCH, as requested by @olebole ).
- deviating from strict standard (e.g. keywords with non-standard ASCII characters beyond upper-case alphanumeric plus dash/underscore) but otherwise unambiguous in intent (i.e., the headers/data are properly described otherwise)
- may be slightly corrupted (e.g. missing end-quote in string values, truncated file or file with terminal junk after)

All in all, I think developers want it to work out of the box, and deal with issues from other people's FITS files automatically. If and when they want to restrict support for common conventions, or want to enforce strict standard, they can still do that as needed.

Specifically, the change would be equivalent to setting:

- `setUseHierarch(true)`
- `setLongStringsEnabled(true)`
- `setAllowHeaderRepairs(true)`
- `setAllowTerminalJunk(true)`

And so the change would be that developers can disable these, instead of enabling them, as needed. Let me know if you think this may not be a good idea after all...

@olebole, @vforchi, @Zlika, @mbtaylor

**Comment by vforchi** [comment]
I don’t have a strong opinion on this, but I think that by default the library should fail if the file is corrupt and it should support the conventions.

**Comment by attipaci** [comment]
For comparison, Two other FITS tools I use a lot, _fv_ (FITS Viewer), and _ds9_ (a FITS image viewer), both open corrupted FITS files by default.That is, they use whatever they can from the FITS. It would be interesting to know what CFITSIO, and PyFITS (astropy.io.fits) do. Do any of you know?

Taking @vforchi 's stance above, one could argue that what applies to _fv_ or _ds9_ is not relevant, since these are compiled tools, which should be the most flexible. 

The counter argument to that, is that if someone tries to read a crappy FITS file with our library, it's because they were handed it as is. It's not like they can go back to where it came from and say, "Hey, you gave me a crappy FITS. Fix it or die.". The only thing you can do with that crappy FITS is try to get whatever is inside it, even if you might not be able to recover all that was intended. At least you get something (and you will never know what was intended anyway...)

**Comment by attipaci** [comment]
BTW, I was just reading the latest FITS standard document (4.0), and the OGIP 1.0 long string convention is not part of the official standard. As such, we should definitely make long string support enabled by default...

**Comment by olebole** [comment]
For a library there are good reasons for both sides, and both seem reasonable. Therefore, I would recommend to keep the default as it is unless there is a really good reason to switch. Otherwise, every dependency would have to check this.

For example, in Debian tamfits is a separate package (libfits-java). Just updating this package would then result in unexpected results, f.e. when a dependency uses the defaults, but wants to be strict (or vice versa). At the end, it is a decision of the application developer what they want. The main purpose here is probably not to have shiny new Java applications, but to keep the library running for their dependencies.

**Comment by attipaci** [comment]
@olebole, one good reason is that the latest FITS standard now specifies long strings as part of the standard (vs an optional convention before), but we currently have it disabled. At minimum we should be able to read/write standard FITS files. Now that means that we need to support long strings out of the box.

So, that alone is a potentially breaking change (as is allowing HIERARCH keywords out of the box, which is what you specifically requested for the next release!). If we do that -- as I think we must -- I'd advocate for the more liberal settings across the board. This way, whatever breaks downstream, has to be fixed once, rather than slow bleeding if we enable features one-by-one. Also, making things more permissive is unlikely to break anything in real life. It _might_ break contrived test cases -- which test for a _specific_ result, rather than a _valid_ result (and I'd argue that in that case the downstream devs really need to fix what they are testing for...)

I know that some Debian packages may rely on nom-tam-fits in their upstream. As such, I propose that we tie the nom-tam-fits release schedule such as to allow Debian developers to properly build and test against the next lib release, and fix whatever needs fixing on their end. I would also support a freeze period for nom-tam-fits before the release (but perhaps excluding critical bug fixes that may be warranted by downstream).

If you let me know what schedule would work best for Debian, and what freeze period before that to allow the Debian developers to make adjustments and test things out. I'd be happy to conform to that....

PS. Please take a look at #169 also, which is another potentially breaking change, that is nonetheless absolutely necessary to fix for scientific use.

**Comment by olebole** [comment]
I think this is not a Debian specific problem; I brought Debian in just as an example here. What I mean is the rule "Don't break compatibility (unless you need to)".
The current default of the library is being strict with respect to the standard, and I think this should be kept. If the standard itself allows more things now, the default of nom-tam-fits should allow the same things by default. I would also think that supporting HIERARCH by default would be OK, as it is an official extension.
I would rather not like to have header repairs or similar enabled by default, however. This would be a significant change in the behaviour which would give surprising results for current users of non-tam-fits (because they may think they can ensure strict compatibility by using the default setting). 
But at the end, this is not that important; and I feel my position here rather as "just my 2 ¢" instead of a strict opposition.
For example, as far as I remember, astropy also changed the FITS file handling behavior from "strict to "try to fix" by default.

**Comment by Zlika** [comment]
I don't think breaking compatibility with older versions is a real problem if you also bump the major version number of the library.
The main feature of this lib is just to read an image file. What would you expect from a function to read a jpeg or png file? Would you expect it to fail (by default) if the file is not strictly compliant with the standard, even if it could still be read?
However, standard validation is also an interesting feature. Maybe this could be fulfilled by a dedicated function that would return a list of all the standard violations detected in the file?

**Comment by attipaci** [comment]
Thanks you all for the input. I'll keep chewing it over, probably right up to the point when it's time to compose the release. I'll almost certainly enable longs string support (because it is standard), and HIERARCH support (because it is a widely accepted convention), but I'm on the fence about the others still. I'm somewhat leaning towards the 'try-to-fix' way (as you could probaly tell), but so far I have not seen any really strong arguments either for or against...

As for the version bump, I think think as long as the API stays backward compatible, a minor bump should suffice. The good old API will stay functional, and the changes in behavior will also be relatively subtle, especially for applications that already use it to compose valid FITS files, or that use the library to read reasonably valid FITS.

**Comment by attipaci** [comment]
PS. @Zlika, the validation I have in mind is mainly for headers you create from a program. For these, I think it's best to throw a runtime exception as soon as you are trying to do something that will result in an invalid or incomplete header. But, I like your idea of keeping track of violations when reading headers (and also when creating, if these violations are not showstoppers to throw an exception, but maybe worth taking note of). I don't think I' prioritize this for the upcoming release (because more thought needs to go into it), but we could slate it for the next release after that (e.g. 1.17)....

**Comment by mbtaylor** [comment]
Being naturally conservative I'm slightly nervous about changes that might modify existing behaviour; but on the whole I agree this change probably makes sense.  One implementation issue: the relevant code in `FitsFactory` for setting `useHierarch` default currently comes with a comment I added for my internal use:
```
        // MBT (28-JUL-2017): change default from false to true.
        // This is required for HIERARCH-based wide fits processing
        // (see uk.ac.starlink.fits.WideFits).  If that gets backed out of,
        // this could be set back to its factory setting (false).
        private boolean useHierarch = true;
```
Whatever the decision on defaults, that comment is not relevant for nom-tam-fits general usage and should be removed.

**Comment by attipaci** [comment]
@wcleveland, I 100% agree with your sentiment, and my goal was/is to preserve support for earlier versions of the standard fully, exactly for the reason you mention. For example, before FITS 4, there was a requirement that ending quotes for string values must not come before byte 20 of the header record. FITS 4 did away with that requirement, but we stick with it anyway, because we want older software to be able to read FITS produced by us. None of the FITS 4 changes otherwise violate any earlier FITS standards. And the choice is yours whether you want to add complex values or hexadecimal integers in headers, which older software may not recognize. (However, we can at least parse them from headers that others produce and we need to handle...)

**Comment by wcleveland** [comment]
I believe the right strategy is require the user to "opt in" to use FITS 4 features, otherwise all the functions will adhere to FITS 3.  Eventually we would release a version that would require the user to "opt out" of FITS 4 to maintain FITS 3 compatibility.

**Comment by attipaci** [comment]
Again, I do not disagree. But, this is where we have been for the past 10 years already with opt-in support things like long strings. As you say, "Eventually we would release a version that would require the user to "opt out" of FITS 4 to maintain FITS 3 compatibility." I would argue now is as good a time as any for that eventuality.

Most of the FITS files (from a number of observatories and instruments) I have been working with in the past 10 years have been using the long string convention, and many have been using the HIERARCH style keyword convention also. In fact FITS4 really just codified what was already widely used in 2016. It's 2021 now. It's time to move beyond opt-in. I honestly find it a nuissance that I have to remember to enable that feature in the library every time I write a tool....

Again, nothing that is added in FITS4 breaks FITS3 compatibility. Any FITS4 can be read as FITS3 just fine. In fact, the standard is careful to evolve in a strictly back compatible way. What FITS4 adds to headers is entirely legal as FITS3, it just has no special meaning in FITS3.

For example, long strings written as FITS4 will read back to a truncated string in FITS3, and the CONTINUE keywords are treated as comment-style keywords.  So, the the enabling long string by default does not change much except adding a few legal comment-style cards from the point of FITS3. Compare that with our <=1.15 behavior with long strings disabled, in which case only the truncated part was added (without comment-style cards after). From the FITS3 point of view the result is essentially the same. 

Same goes for HIERARCH keywords. These are just comment style cards as far as FITS3 is concerned. They just have no meaning when the convention is disabled. So we can safely write these by default, and a FITS3 reader will just treat them as comments. Compare that with the behavior of `setUseHierarch()` disabled by default (<=1.15), where these keys were simply not written to the header. The result is again essentially the same from a FITS3 point of view...

Same goes for complex header values. (There isn't even a switch for that.) Before I added support Headers containing complex values cards stll have been parsed as a non-string value, which could be accessed as a string. With FITS3 it would be up to your application to make sense of that string. Now you can access it as a complex value more conveniently. But nothing really changes except that we added convenience methods for accessing it.

I'm quite confident that permissive settings is extremely unlikely to break any legacy code. That is not to say that 1.16 won't break existing applications, but that is far more likely to happen as a result of bugs that were fixed, and which legacy code may have worked around in creative ways...

**Comment by attipaci** [comment]
And as a separate note, I also think we should not be too afraid of breaking stuff if it is for a good reason. This library has done it before, as do others (Python 3 anyone, lol?). I had to fix up some of my older code to work with newer releases of this library in the past. And that was fine, because the newer releases were generally an improvement over the old ones... So, yes, we should tread carefully, but we should not hesitate to make things better just because we might lose a sliver of 100% back compatibility...

**Comment by olebole** [comment]
I would also recommend to get in contact with the Astropy fits package maintainer, @saimn. IMO it would be a good idea to have tamfits and astropy interoperable by default. Also, he may have some arguments here from the experiences with Astropy.
For future development, it may be useful to keep coordination of the default settings, and also maybe to have some inter-language CI tests.

**Comment by attipaci** [comment]
You can now find a pre-release (1.16.0-rc1) under the Releases section on the main page, with a precompiled (unsigned) JAR. Or, feel free to build your own from __master__.

Give it a go for your own testing and applications. From now until the 1.16.0 release I plan to only address fixing blocking issues. Depending on how the testing goes and how many issues need to be fixed, the official 1.16.0 release may come in December 2021 or January 2022 (and hopefully not much beyond). I made an effort to maintain back compatibility as far as possible. If something slipped up, let me know. (I'll start a new ticket for 1.16.0 pre-releases...)

I look forward to your feedback, and learning about any problems you might encounter. Thanks a lot for your help (past, present, and future)

-- A.
